### PR TITLE
SHOW VITESS_MIGRATION '...' LOGS, retain logs for 24 hours

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description
Fixes https://github.com/vitessio/vitess/issues/8178

- Support for `show vitess_migration '<uuid>' logs` queries, e.g. `show vitess_migration '4835ca51_e867_11eb_ab12_0a43f95f28a3' logs`
- Retaining migration logs for 24 hours after migration completion, whether migration is successful or not.

`gh-ost` and `pt-osc` (but not `online`) migrations generate a migration log, which Vitess stores in a well known location on the primary tablet.

It is now possible to view the file, during or after migration. The file is retained for 24 hours after completion/failure.

The command `show vitess_migration '<uuid>' logs` will output the entire log file content for each shard. Therefore, 4 shards => 4 log files.

This is tested in `go/test/endtoend/onlineddl/ghost/onlineddl_ghost_test.go`

Documentation to follow

Example:
```
mysql> show vitess_migration '3a273866_e867_11eb_ab12_0a43f95f28a3' logs \G
*************************** 1. row ***************************
migration_log: 2021-07-19 07:59:23 INFO starting gh-ost 261355426d8fc31b590733ca8ff8e79012103c18
2021-07-19 07:59:23 INFO Migrating `vt_commerce`.`corder`
2021-07-19 07:59:23 INFO executing gh-ost-on-startup hook: /tmp/online-ddl-3a273866_e867_11eb_ab12_0a43f95f28a3-943208852/gh-ost-on-startup
ok
2021-07-19 07:59:23 INFO inspector connection validated on ip-REDACTED:17100
2021-07-19 07:59:23 INFO User has SUPER, REPLICATION SLAVE privileges, and has ALL privileges on `vt_commerce`.*
2021-07-19 07:59:23 INFO binary logs validated on ip-REDACTED:17100
2021-07-19 07:59:23 INFO Restarting replication on ip-REDACTED:17100 to make sure binlog settings apply to replication thread
2021-07-19 07:59:23 INFO Inspector initiated on ip-REDACTED:17100, version 5.7.30-log
2021-07-19 07:59:23 INFO Table found. Engine=InnoDB
...
```

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/8178
- #6926 

## Checklist
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/engine/show_migration_logs.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	"vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+var _ Primitive = (*ShowMigrationLogs)(nil)
+
+//ShowMigrationLogs represents the instructions to perform an online schema change via vtctld
```
Comment (systay) [inline go/vt/vtgate/engine/show_migration_logs.go:31]: nit: we should have a space after //

File: go/vt/vtgate/engine/show_migration_logs.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// Execute implements the Primitive interface
+func (v *ShowMigrationLogs) Execute(vcursor VCursor, bindVars map[string]*query.BindVariable, wantfields bool) (result *sqltypes.Result, err error) {
+	s := Send{
+		Keyspace:          v.Keyspace,
+		TargetDestination: v.TargetDestination,
+		Query:             v.Query,
+		IsDML:             false,
+		SingleShardOnly:   false,
+	}
+	result, err = s.Execute(vcursor, bindVars, wantfields)
```
Comment (systay) [inline go/vt/vtgate/engine/show_migration_logs.go:77]: nit: can return directly from `s.Execute`

File: go/vt/vtgate/engine/show_migration_logs.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vtgate/engine/show_migration_logs.go]: fixed

File: go/vt/vtgate/engine/show_migration_logs.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vtgate/engine/show_migration_logs.go]: fixed

File: go/vt/vtgate/engine/show_migration_logs.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	return v.Keyspace.Name
+}
+
+// GetTableName implements the Primitive interface
+func (v *ShowMigrationLogs) GetTableName() string {
+	return ""
+}
+
+// Execute implements the Primitive interface
+func (v *ShowMigrationLogs) Execute(vcursor VCursor, bindVars map[string]*query.BindVariable, wantfields bool) (result *sqltypes.Result, err error) {
```
Comment (systay) [inline go/vt/vtgate/engine/show_migration_logs.go:69]: Why add this primitive instead of producing the `Send` in the first place?

Comment (shlomi-noach) [reply go/vt/vtgate/engine/show_migration_logs.go]: Resolved. Now using plain `Send`.

**Comment by shlomi-noach** [comment]
This PR has been backported to release-11 in https://github.com/vitessio/vitess/pull/8532. If I understand correctly, we will merge release-11 into `main`, so _this_ PR is no longer necessary, right?

**Comment by shlomi-noach** [comment]
ping for review :pray:

**Review by systay — APPROVED** [review]

**Comment by shlomi-noach** [comment]
> This PR has been backported to release-11 in #8532. If I understand correctly, we will merge release-11 into main, so this PR is no longer necessary, right?

Self answering: nope.

**Pull request merged**